tests/functional/test_vcs_git.py

```diff
@@ -282,3 +278,66 @@ def test_resolve_commit_not_on_branch(script, tmp_path):
     # check we can fetch our commit
     rev_options = Git.make_rev_options(commit)
     Git().fetch_new(str(clone_path), repo_path.as_uri(), rev_options)
+
+
+def test_fetch_new(script, tmp_path):
+    repo_path = tmp_path / "repo"
+    repo_file = repo_path / "file.txt"
+    clone_path1 = repo_path / "clone1"
+    clone_path2 = repo_path / "clone2"
+    clone_path3 = repo_path / "clone3"
+
+    repo_path.mkdir()
+    script.run("git", "init", cwd=str(repo_path))
+    repo_file.write_text(u".")
+    script.run("git", "add", "file.txt", cwd=str(repo_path))
+    script.run("git", "commit", "-m", "initial commit", cwd=str(repo_path))
+    commit = script.run("git", "rev-parse", "HEAD", cwd=str(repo_path)).stdout.strip()
+
+    # Check that we can clone at HEAD
+    Git().fetch_new(str(clone_path1), repo_path.as_uri(), Git.make_rev_options())
+    # Check that we can clone to commit
+    Git().fetch_new(str(clone_path2), repo_path.as_uri(), Git.make_rev_options(commit))
+    # Check that we can clone w/ old version of git w/o --filter
+    with patch("pip._internal.vcs.git.Git.get_git_version", return_value=(2, 16)):
```

Comment:
this is not quite as good as actually installing an old git in the CI environment, but it's probably good enough for testing - just making sure the "else" codepath is functional. I manually tested with an actual old git.

Reply:
Agreed, goo enough for testing. I would actually split this into two tests, one patching to return `(2, 16)` and one skipped if the Git version is lower than 2.17, so a contributor running Git < 2.17 would still be able to make the test suite pass.

Reply:
This entire test case should be split into several smaller ones.

Reply:
Split into three cases
1) Supported client, supported server
2) Supported client, unsupported server
3) Unsupported client (no --filter)